pkg/storage/watch_cache.go

```diff
@@ -250,6 +250,20 @@ func (w *watchCache) GetAllEventsSinceThreadUnsafe(resourceVersion uint64) ([]wa
 	if size > 0 {
 		oldest = w.cache[w.startIndex%w.capacity].resourceVersion
 	}
+	if resourceVersion == 0 {
+		// resourceVersion = 0 means that we don't require any specific starting point
+		// and we would like to start watching from ~now.
+		// However, to keep backward compatibility, we additionally need to return the
+		// current state and only then start watching from that point.
+		//
+		// TODO: In v2 api, we should stop returning the current state - #13969.
+		allItems := w.store.List()
```

Comment:
Wait, why is this feature not safe?  The original implementation was safe -
we called etcd to do a list, then watched from the list timestamp.  How is
the new version less able to deliver safety?

On Tue, Sep 15, 2015 at 1:30 PM, Daniel Smith <notifications@github.com>
wrote:

> In pkg/storage/watch_cache.go
> <https://github.com/kubernetes/kubernetes/pull/13910#discussion_r39539558>
> :
>
> > @@ -250,6 +250,20 @@ func (w *watchCache) GetAllEventsSinceThreadUnsafe(resourceVersion uint64) ([]wa
> >  	if size > 0 {
> >  		oldest = w.cache[w.startIndex%w.capacity].resourceVersion
> >  	}
> > +	if resourceVersion == 0 {
> > +		// resourceVersion = 0 means that we don't require any specific starting point
> > +		// and we would like to start watching from ~now.
> > +		// However, to keep backward compatibility, we additionally need to return the
> > +		// current state and only then start watching from that point.
> > +		//
> > +		// TODO: In v2 api, we should stop returning the current state - #13969.
> > +		allItems := w.store.List()
>
> Does this list go out to hit etcd? I guess it doesn't matter since this
> feature can't be used safely anyway, but if it doesn't go out to hit etcd
> there's no guarantee the combined state returned by the list would be a
> complete/correct one as of any particular resource version.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13910/files#r39539558>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift